### PR TITLE
[doc] Update tutorials to the new sanity and performance syntax

### LINKED ISSUE #2082: Adapt the tutorial examples to use new `sanity_function` built-in
I'd suggest to remove the not so natural redirection to `hello.out` and go for the sanity function. We don't even need to mention anything about the pipeline now.

### PR DESCRIPTION
**Pull request by jjotero**
Closes #2082

### CODE REVIEW

File: docs/tutorial_basics.rst [hunk 1]
```diff
@@ -469,15 +467,8 @@ In this case, we set the :func:`set_compile_flags` hook to run before the ReFram
...
-Similarly to the parameters and the compiler settings, the output of a test is private to each of the instances of the :class:`HelloThreadedTest` class.
-So, instead of inspecting an external file to evaluate the sanity of the test, we can just set our sanity function to inspect this attribute that contains the test's standard output.
-This output is stored under :attr:`self.stdout` and is populated only after the executable has run.
-Therefore, we can set the :attr:`~reframe.core.pipeline.RegressionTest.sanity_patterns` with the :func:`set_sanity_patterns` pipeline hook that is scheduled to run before the ``sanity`` pipeline stage.
-Again, pipeline stages will be covered detail further on, so for now, just think of this ``sanity`` stage as a step that occurs after the test's executable is run.
+In this example, the generated executable takes a single argument which sets the number of threads to be used.
+The options passed to the test's executable can be set throught the :attr:`executable_opts <reframe.core.pipeline.RegressionTest.executable_opts>` variable, which in this case is set to ``'16'``.
```
Comment (jgphpc) [inline docs/tutorial_basics.rst:471]:
```suggestion
The options passed to the test's executable can be set through the :attr:`executable_opts <reframe.core.pipeline.RegressionTest.executable_opts>` variable, which in this case is set to ``'16'``.
```

File: docs/tutorial_advanced.rst [hunk 2]
```diff
@@ -625,7 +625,7 @@ It resembles a scaling test, except that all happens inside a single ReFrame tes
 
 .. literalinclude:: ../tutorials/advanced/multilaunch/multilaunch.py
    :lines: 6-
-   :emphasize-lines: 12-19
+   :emphasize-lines: 13-19
```
Comment (jgphpc) [inline docs/tutorial_advanced.rst:628]: typo @line 495: `not use Slurm  and and another option`

File: docs/tutorial_tips_tricks.rst [hunk 3]
```diff
@@ -146,15 +146,15 @@ If we run the test, we can see that the correct standard output filename will be
 
 Debugging sanity and performance patterns
 =========================================
-When creating a new test that requires a complex output parsing for either the sanity or performance stages, setting the :attr:`sanity_patterns` and :attr:`perf_patterns` may involve some trial and error to debug the complex regular expressions required.
+When creating a new test that requires a complex output parsing for either the ``sanity`` or ``performance`` pipeline stages, tunning the functions decorated by :attr:`@sanity_function<reframe.core.pipeline.RegressionMixin.sanity_function>` or :attr:`@performance_function<reframe.core.pipeline.RegressionMixin.performance_function>` may involve some trial and error to debug the complex regular expressions required.
```
Comment (jgphpc) [inline docs/tutorial_tips_tricks.rst:149]:
```suggestion
When creating a new test that requires a complex output parsing for either the ``sanity`` or ``performance`` pipeline stages, tuning the functions decorated by :attr:`@sanity_function<reframe.core.pipeline.RegressionMixin.sanity_function>` or :attr:`@performance_function<reframe.core.pipeline.RegressionMixin.performance_function>` may involve some trial and error to debug the complex regular expressions required.
```

File: docs/tutorial_basics.rst (hunk 1 above)
Comment (jgphpc) [reply docs/tutorial_basics.rst:471]: While `attr:executable_opts ...` link is working fine here, it does not above (line 361)

File: docs/tutorial_basics.rst [hunk 4]
```diff
@@ -789,6 +779,28 @@ If any obtained performance value is beyond its respective thresholds, the test
...
+
+      @run_before('performance')
+      def set_perf_vars(self):
+          self.perf_variables = {
+              'Copy': sn.make_performance_function(
+                  sn.extractsingle(r'Copy:\s+(\S+)\s+.*',
+                                   self.stdout, 1, float),
+                  'MB/s'
+               )
+          }
```
Comment (jgphpc) [inline docs/tutorial_basics.rst:803]: typo@line 812: and `at` the same time ?

File: docs/tutorial_advanced.rst [hunk 5]
```diff
@@ -281,7 +281,7 @@ The following test is a compile-only version of the :class:`MakefileTest` presen
 
 
 .. literalinclude:: ../tutorials/advanced/makefiles/maketest.py
-   :lines: 32-
+   :lines: 30-
```
Comment (jgphpc) [inline docs/tutorial_advanced.rst:284]: @line 167: `The compiler variables ...` should appear in the first box ?

Comment (jgphpc) [reply docs/tutorial_advanced.rst:284]: @line 253: lower and upper bounds`.`

File: docs/tutorial_basics.rst (hunk 1 above)
Comment (jjotero) [reply docs/tutorial_basics.rst:471]: Yeah, I'm not sure you can do that. I think the only solution would be to have a link to the config page.

File: docs/tutorial_advanced.rst (hunk 5 above)
Comment (jjotero) [reply docs/tutorial_advanced.rst:284]: Not sure what you mean on the first comment.

Comment (jgphpc) [reply docs/tutorial_advanced.rst:284]: I would expect to see:
`make -j 1 CPPFLAGS="-DELEM_TYPE=float" CC=cc CXX=CC` 
?

File: docs/tutorial_basics.rst [hunk 6]
```diff
@@ -789,6 +779,28 @@ If any obtained performance value is beyond its respective thresholds, the test
      * Rerun with '-n StreamWithRefTest -p gnu --system catalina:default'
      * Reason: performance error: failed to meet reference: Copy=24586.5, expected 55200 (l=52440.0, u=57960.0)
 
+Also, note how the performance syntax for this example is far more compact in comparison to our first iteration of the STREAM test.
```
Comment (vkarak) [inline docs/tutorial_basics.rst:782]: I would provide that in a separate version of the stream test (so as to be able to test it as well). At the beginning of the paragraph we are talking only about the reference and here we show a completely different implementation of the performance variables extraction. I would add all this to a subsection named perhaps "setting explicitly the test's performance variables."

File: docs/tutorial_basics.rst [hunk 7]
```diff
@@ -789,6 +779,28 @@ If any obtained performance value is beyond its respective thresholds, the test
...
+In that first STREAM example, all four performance functions were almost identical, except for a small part of the regex pattern, which led to a lot of code repetition.
+Hence, this example collapses all four performance functions into a single performance function, which now takes an optional argument to select the quantity to extract.
+Then, the performance variables of the test can be defined by setting the respective entries in the :attr:`~reframe.core.pipeline.RegressionTest.perf_variables` dictionary.
+
+.. literalinclude:: ../tutorials/basics/stream/stream2.py
+   :lines: 41-
+
+.. note::
```
Comment (vkarak) [inline docs/tutorial_basics.rst:790]:
```suggestion
.. tip::
```

File: docs/tutorial_tips_tricks.rst [hunk 8]
```diff
@@ -90,7 +90,7 @@ As suggested by the warning message, passing :option:`-v` will give you the stac
 Debugging deferred expressions
 ==============================
 
-Although deferred expression that are used in :attr:`sanity_patterns` and :attr:`perf_patterns` behave similarly to normal Python expressions, you need to understand their `implicit evaluation rules <deferrable_functions_reference.html#implicit-evaluation-of-sanity-functions>`__.
+Although deferred expression that are used in sanity and performance functions behave similarly to normal Python expressions, you need to understand their `implicit evaluation rules <deferrable_functions_reference.html#implicit-evaluation-of-sanity-functions>`__.
```
Comment (vkarak) [inline docs/tutorial_tips_tricks.rst:93]:
```suggestion
Although deferred expressions that are used in sanity and performance functions behave similarly to normal Python expressions, you need to understand their `implicit evaluation rules <deferrable_functions_reference.html#implicit-evaluation-of-sanity-functions>`__.
```

File: docs/tutorial_advanced.rst (hunk 5 above)
Comment (vkarak) [reply docs/tutorial_advanced.rst:284]: @jgphpc You are right here. @jjotero he means that the emitted line is not correct as we show it in the tutorial.

File: docs/tutorial_advanced.rst [hunk 9]
```diff
@@ -680,13 +680,9 @@ The test will verify that all the nodes print the expected host name:
...
 
-The first thing to notice in this test is that :attr:`~reframe.core.pipeline.RegressionTest.num_tasks` is set to zero.
-This is a requirement for flexible tests.
-The sanity check of this test simply counts the host names printed and verifies that they are as many as expected.
-Notice, however, that the sanity check does not use :attr:`num_tasks` directly, but rather access the attribute through the :func:`~reframe.utility.sanity.getattr` sanity function, which is a replacement for the :func:`getattr` builtin.
-The reason for that is that at the time the sanity check expression is created, :attr:`num_tasks` is ``0`` and it will only be set to its actual value during the run phase.
-Consequently, we need to defer the attribute retrieval, thus we use the :func:`~reframe.utility.sanity.getattr` sanity function instead of accessing it directly
-
+The first thing to notice in this test is that :attr:`~reframe.core.pipeline.RegressionTest.num_tasks` is set to zero as default, which is a requirement for flexible tests.
+However, this value is set to the actual number of tasks during the ``run`` pipeline stage.
```
Comment (vkarak) [inline docs/tutorial_advanced.rst:684]: I would move this after the following sentence (starting without the "lastly") and rephrase it to something that the number of tasks is set after the job finishes to the actual number of tasks that were used.

**Review by vkarak — CHANGES_REQUESTED** [review]
That was a lot of work! I have some comments for improvements and if we want to be very pedantic, we should rerun the tutorial examples and update the output wherever we list it, so that the ReFrame version in the output is meaningful.

File: docs/tutorial_basics.rst [hunk 10]
```diff
@@ -678,9 +663,13 @@ In this case, we just fetch the source code of the benchmark.
...
-In this example, we extract four performance variables, namely the memory bandwidth values for each of the "Copy", "Scale", "Add" and "Triad" sub-benchmarks of STREAM and we do so by using the :func:`~reframe.utility.sanity.extractsingle` sanity function.
+What makes a ReFrame test a performance test is the definition of at least one :ref:`performance function<deferrable-performance-functions>`.
+Similarly to a test's :func:`@sanity_function<reframe.core.pipeline.RegressionMixin.sanity_function>`, a performance function is simply a member function decorated with the :attr:`@performance_function<reframe.core.pipeline.RegressionMixin.performance_function>` decorator, which is responsible for extracting a specified performance quantity from a regression test.
+The :attr:`@performance_function<reframe.core.pipeline.RegressionMixin.performance_function>` decorator must be passed the units of the quantity to be extracted, and it also takes the optional argument ``perf_key`` to customize the name of the extracted performance variable.
+If ``perf_key`` is not provided, the performance variable will take the name of the decorated performance function.
```
[the remark this answers is not in the document]
Comment (jjotero) [reply docs/tutorial_basics.rst:669]: Hmm, a performance function does not necessarily define a performance variable. Such a sentence could be interpreted as that one needs a performance function for each performance variable, but that is not true. I have reworded all this section here, and it is now hopefully a bit more clear.

File: docs/tutorial_basics.rst [hunk 11]
```diff
@@ -678,9 +663,13 @@ In this case, we just fetch the source code of the benchmark.
...
-In this example, we extract four performance variables, namely the memory bandwidth values for each of the "Copy", "Scale", "Add" and "Triad" sub-benchmarks of STREAM and we do so by using the :func:`~reframe.utility.sanity.extractsingle` sanity function.
+What makes a ReFrame test a performance test is the definition of at least one :ref:`performance function<deferrable-performance-functions>`.
+Similarly to a test's :func:`@sanity_function<reframe.core.pipeline.RegressionMixin.sanity_function>`, a performance function is simply a member function decorated with the :attr:`@performance_function<reframe.core.pipeline.RegressionMixin.performance_function>` decorator, which is responsible for extracting a specified performance quantity from a regression test.
+The :attr:`@performance_function<reframe.core.pipeline.RegressionMixin.performance_function>` decorator must be passed the units of the quantity to be extracted, and it also takes the optional argument ``perf_key`` to customize the name of the extracted performance variable.
```
[the remark this answers is not in the document]
Comment (jjotero) [reply docs/tutorial_basics.rst:668]: I have reworded this whole section for clarity.

File: docs/tutorial_advanced.rst (hunk 9 above)
Comment (jjotero) [reply docs/tutorial_advanced.rst:684]: I'm not sure that breaking the chronological order of things would make this more clear. I've reworded this a bit.

File: docs/tutorial_basics.rst (hunk 6 above)
Comment (jjotero) [reply docs/tutorial_basics.rst]: Makes sense 👍

File: docs/tutorial_basics.rst (hunk 10 above)
[the remark this answers is not in the document]
Comment (vkarak) [reply docs/tutorial_basics.rst]: Yes, you are right.

File: tutorials/basics/stream/stream2.py [hunk 12]
```diff
@@ -20,14 +20,6 @@ class StreamWithRefTest(rfm.RegressionTest):
...
         'OMP_PLACES': 'cores'
     }
-    reference = {
-        'catalina': {
-            'Copy':  (25200, -0.05, 0.05, 'MB/s'),
-            'Scale': (16800, -0.05, 0.05, 'MB/s'),
-            'Add':   (18500, -0.05, 0.05, 'MB/s'),
-            'Triad': (18800, -0.05, 0.05, 'MB/s')
-        }
-    }
```
Comment (vkarak) [inline tutorials/basics/stream/stream2.py:30]: Since we removed the reference here, we should probably rename the test as well.

File: docs/tutorial_basics.rst [hunk 13]
```diff
@@ -1122,9 +1140,9 @@ Let's see and comment the changes:
 
 .. code-block:: console
 
-   cat tutorials/basics/stream/stream3.py
+   cat tutorials/basics/stream/stream4.py
 
-.. literalinclude:: ../tutorials/basics/stream/stream3.py
+.. literalinclude:: ../tutorials/basics/stream/stream4.py
```
Comment (vkarak) [inline docs/tutorial_basics.rst:1145]: I think you forgot to `git add` this.

File: docs/tutorial_advanced.rst [hunk 14]
```diff
@@ -161,7 +161,7 @@ Let's inspect the build script generated by ReFrame:
 
    trap _onerror ERR
 
-   make -j 1 CPPFLAGS="-DELEM_TYPE=float"
+   make -j 1 CPPFLAGS="-DELEM_TYPE=float" CC=cc CXX=CC
```
Comment (vkarak) [inline docs/tutorial_advanced.rst:164]: The line is actually this:

```suggestion
   make -j 1 CC="cc" CXX="CC" FC="ftn" NVCC="nvcc" CPPFLAGS="-DELEM_TYPE=float"
```

**Review by vkarak — CHANGES_REQUESTED** [review]
Just a couple of additional minor comments.

**Comment by vkarak** [comment]
Looks good. Only thing remaining is to add the `stream4.py`.

**Comment by jjotero** [comment]
@vkarak I just added the missing file for the `stream4.py` tutorial. Thanks for addressing the other stuff!

**Review by vkarak — APPROVED** [review]
